### PR TITLE
Fix #1497-BackButton Fix

### LINKED ISSUE #1497: Backward button is clickable on first question in form if user changes device orientation
#### Software and hardware versions 
Collect - current master
Android 4.1, 4.2, 4.4, 5.1, 6.0, 7.0

#### Problem description
Backward button is clickable on first question in form if user changes device orientation

#### Steps to reproduce the problem
1. User uses forward/backward buttons to navigate
2. User is on first question in form
3. Backward button is visible but it is not clickable
4. User changes device orientation
5. Backward button is clickable

#### Expected behavior
Backward button is not clickable

### PR DESCRIPTION
**Pull request by codedsun**
Closes #1497 

#### What has been done to verify that this works as intended?
Followed the steps which were causing the issue and found that issue was fixed.

**Edit 1**
I checked for the following
1.onRotation the showNavigationButtons value is fixed.
2.onResume the issue is fixed too!
3.After onStop following onDestroy() the issue is fixed too.

#### Why is this the best possible solution? Were any other approaches considered?
Because on Orientation changes, onSavedInstanceState and onRestoreInstanceState are called and hence the method to adjust backward button visibility is called there.

#### Are there any risks to merging this code? If so, what are they?
No

#### Do we need any specific form for testing your changes? If so, please attach one.
No

### CODE REVIEW

**Comment by codedsun** [comment]
@grzesiek2010 @mmarciniak90 Kindly review!

**Comment by grzesiek2010** [comment]
@codedsun 
Congrats on your first pull request!

When it comes to the code your solution works as expected but it's not appropriate. I'll try to explain why:
> Because on Orientation changes, onSavedInstanceState and onRestoreInstanceState are called

You are right `onSavedInstanceState()` is called before recreation and `onRestoreInstanceState()` is called after it. 
The first thing is you don't need to call `adjustBackNavigationButtonVisibility()` in `onSavedInstanceState()` since it's before recreation and it doesn't make sense. 
Calling that method in `onRestoreInstanceState()` would be enough but it's not needed too. Why? Because `onCreate()` method (where `adjustBackNavigationButtonVisibility()` is called originally) is called after recreation as well. The problem is that `showNavigationButtons` field [here](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java#L1346) has a false value if you change your device orientation (or do something else what causes recreation) and it blocks calling `adjustBackNavigationButtonVisibility()`. That value should be true of course since we want to see buttons.

So the problem is that `showNavigationButtons` is true only just after opening a form and then if you eg. change your device orientation it's false what's wrong.

**Comment by grzesiek2010** [comment]
@opendatakit-bot unlabel "needs testing"

**Comment by grzesiek2010** [comment]
@opendatakit-bot unlabel "needs review"

**Comment by grzesiek2010** [comment]
@opendatakit-bot label "changes requested"

**Comment by grzesiek2010** [comment]
...and you have a wrong issue number in your description after `Closes #`.

**Comment by codedsun** [comment]
@grzesiek2010  as you stated the adjustBackNavigationButtonVisibility() is called in the onCreate ( ) method ! but I logged and checked
The logcat shows with this PR is 
10-06 22:04:23.848 21455-21455/org.odk.collect.android E/TAG: onSaveInstanceState: 
10-06 22:04:23.849 21455-21455/org.odk.collect.android E/TAG: adjustBackNavigationButtonVisibility: 
10-06 22:04:23.932 21455-21455/org.odk.collect.android E/TAG: onCreate: 
10-06 22:04:24.014 21455-21455/org.odk.collect.android E/TAG: onRestoreInstanceState: 
10-06 22:04:24.014 21455-21455/org.odk.collect.android E/TAG: adjustBackNavigationButtonVisibility: 

This means that adjustBackNavigationButtonVisibility is not called in onCreate, it 's called in createView ()
Tell me where I am going wrong , I am also newbie !

**Comment by grzesiek2010** [comment]
@codedsun sorry my mistake I meant `createView()` of course not `onCreate()` (probably I used `onCreate()` because everything starts from that method... never mind).
So everything I said above is correct but I was talking about `createView()` method.
Just to recap. This method is called again after recreation but adjustBackNavigationButtonVisibility() is not called because of https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java#L1346 and this is a problem - `showNavigationButtons` is false but it should be true.

**Comment by codedsun** [comment]
@grzesiek2010  Ok ! let me dive more into it and find what's the problem.

**Comment by grzesiek2010** [comment]
@codedsun you can use debugging instead of logs its easier https://developer.android.com/studio/debug/index.html

**Comment by codedsun** [comment]
@grzesiek2010 I think the issue is when the app starts, the onResume method is called before the CreateView( ) while when we change the orientation the CreateView() is called before the onResume () hence the value of showNavigationButton is false when the orienation changes and still when you move from landscape to portrait the showNavigationButton is still false
Guide me for the same.

**Comment by codedsun** [comment]
@grzesiek2010 the check  
if (navigation.contains(PreferenceKeys.NAVIGATION_BUTTONS)) {
            showNavigationButtons = true;
        }

should made in createView() and will solve the issue .

**Comment by grzesiek2010** [comment]
@codedsun yes this is a problem. This fragment of code:

> if (navigation.contains(PreferenceKeys.NAVIGATION_BUTTONS)) {
showNavigationButtons = true;
}

should be called before creating layout since it's an important value for us.

**Comment by codedsun** [comment]
Ok will be sending PR by today.

**Comment by codedsun** [comment]
@grzesiek2010 It have a problem .

```
// only check the buttons if it's enabled in preferences
        String navigation = (String) GeneralSharedPreferences.getInstance().get(PreferenceKeys.KEY_NAVIGATION);
        if (navigation.contains(PreferenceKeys.NAVIGATION_BUTTONS)) {
            showNavigationButtons = true;
        }

```
When this code is pasted in createView() and remove from onResume () , first time its working fine.
but when i move back and again come to the main menu (without closing the app) and then again move to the section of fill bank form the value of show navigation button is true still they are not shown 

And when the code is pasted in createView() and not removed from onResume () the above problem vanishes and code is working fine. 

So what should I do ?

**Comment by grzesiek2010** [comment]
@codedsun 
I've already said when you asked my on Slack that adding anything in createView() is a wrong approach since createView() is called every time you navigate through widgets and we don't need that - we need to adjust buttons when the activity is created, only once. So it must be in onCreate() or in onResume() but before you call that createView() for the first time.

> And when the code is pasted in createView() and not removed from onResume () the above problem vanishes and code is working fine.
So what should I do ?

I believe it works for you but I also believe we can adjust buttons only in one place (maybe I'm wrong it's your job to check it debugging the project).

**Comment by grzesiek2010** [comment]
@codedsun and I think you missed my advice:

> The problem is that showNavigationButtons field here has a false value if you change your device orientation (or do something else what causes recreation) and it blocks calling adjustBackNavigationButtonVisibility(). That value should be true of course since we want to see buttons.

it's above and it's the main problem. You don't need to adjust buttons in another place just fix the issue with showNavigationButtons field and it should work. Am I wrong?

**Comment by codedsun** [comment]
I logged some methods and saw that  I set the in onSavedInstanceState method the shownavigationButtons  to true 
but still in the createView they are set to false.

[6:44] 
10-16 18:43:10.498 8264-8264/org.odk.collect.android E/TAG: onSaveInstanceState: trueNAVIGATION buttons
10-16 18:43:10.639 8264-8264/org.odk.collect.android E/TAG: createView: false
10-16 18:43:10.646 8264-8264/org.odk.collect.android E/TAG: onRestoreInstanceState: 
10-16 18:43:10.725 8264-8264/org.odk.collect.android E/TAG: createView: false
10-16 18:43:10.727 8264-8264/org.odk.collect.android E/TAG: onResume: true

**Comment by grzesiek2010** [comment]
@codedsun 
probably `createView()` is called before you read `shownavigationButtons` value after activity recreation or maybe you don't read that value at all.

**Comment by grzesiek2010** [comment]
@codedsun 
Looks like you don't understand [Activity lifecycle](https://developer.android.com/guide/components/activities/activity-lifecycle.html) in Android (no worries you are new as I know). 

Everything you need to solve the problem is a knowledge about mentioned lifecycle and about [debugging android projects](https://developer.android.com/studio/debug/index.html) (using logs/tags works too but it takes more time).

When you open a new form it works well because `shownavigationButtons` is set in `onResume()` and `createView()` is called from [loadingComplete()](https://github.com/opendatakit/collect/blob/master/collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java#L2687) (not directly it starts from `refreshCurrentView()`).

in that case `onResume()` is called before because your device needs some time to load the form. But then if you change your device orientation (to force recreation) `createView()` is called before (since we don't load the form again and `createView()` is called somewhere in `onCreate()` probably).

So the problem is we set `shownavigationButtons` in `onResume()` and it's too late it should be somewhere in `onCreate()`. And probably that's all unless there is any other strange case I can't imagine.

I can't be more precise :smile: I would recommend you to choose another issue but this is a great one for new developers and on the other hand we don't have many easier issues than this one.

**Comment by codedsun** [comment]
@grzesiek2010 PR Updated!

File: collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java [hunk 1]
```diff
@@ -1368,6 +1373,11 @@ public void onClick(View v) {
         }
     }
 
+    @Override
+    protected void onRestoreInstanceState(Bundle savedInstanceState) {
```
Comment (grzesiek2010) [inline collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java]: you don't need this.

Comment (codedsun) [reply collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java]: done updated !

**Comment by codedsun** [comment]
**What has been done to verify that this works as intended?**

For the following issue ,I checked for the following

1. onRotation the showNavigationButtons value is fixed.
2.onResume the issue is fixed too! 
3.After onStop following onDestroy() the issue is fixed too.

**Comment by grzesiek2010** [comment]
Ok at first glance looks ok, but I wonder why someone put that code in `onResume()` before so maybe some strange case I can't imagine exist. @mmarciniak90 please test it carefully.

@opendatakit-bot unlabel "changes requested"
@opendatakit-bot label "needs testing"

**Comment by mmarciniak90** [comment]
Tested with success.
Tested on Android 4.1, 4.2, 4.4, 5.1, 6.0, 7.0
Backward button is not clickable on first question in form even if user changes device orientation.
I also verified locking the screen, minimizing the app and behavior with turned on 'Don't keep activities' option.

@opendatakit-bot unlabel "needs testing"

**Review by lognaturel — APPROVED** [review]

**Comment by lognaturel** [comment]
Thank you so much for this fix @codedsun and for taking the time to iterate with @grzesiek2010 and get to the simplest possible solution.

Welcome to ODK! 😊 If you haven't already, make sure to join the developer Slack at http://slack.opendatakit.org and to check out the forum at https://forum.opendatakit.org

**Pull request merged**